### PR TITLE
Feature/ml tidy

### PR DESCRIPTION
**Pull request by baykaner**

### CODE REVIEW

File: libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp [hunk 1]
```diff
@@ -474,6 +474,7 @@ void ReduceSum(ArrayType const &obj1, typename ArrayType::SizeType axis, ArrayTy
       {
         ret[i] += obj1({j, i});
       }
+      std::cout << "ret[i]: " << ret[i] << std::endl;
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/matrix_operations/matrix_operations.hpp]: Debug code to remove

File: libs/math/include/math/free_functions/ml/activation_functions/relu.hpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+template <typename ArrayType>
+ArrayType Relu(ArrayType &t)
+{
+  ArrayType ret;
+  ret.Copy(t);
+  Relu(ret, ret);
+  return ret;
+}
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/ml/activation_functions/relu.hpp:47]: {
   ArrayType ret(t.shape());
   Relu(t, ret);
   return ret; 
}

No need for a copy here since Relu L32 overwrite systematically.

File: libs/math/include/math/free_functions/ml/loss_functions/l2_norm.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+namespace math {
+
+template <typename ArrayType>
+typename ArrayType::Type L2Norm(ArrayType const &A, ArrayType &ret)
+{
+  assert(A.size() == ret.size());
+  assert(A.shape() == ret.shape());
+
+  Square(A, ret);
+  return std::sqrt(Sum(ret));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/ml/loss_functions/l2_norm.hpp:35]: This will likely not compile with fixed point

File: libs/math/include/math/free_functions/ml/loss_functions/mean_square_error.hpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+template <typename ArrayType>
+typename ArrayType::Type MeanSquareError(ArrayType const &A, ArrayType const &B)
+{
+  assert(A.shape() == B.shape());
+  ArrayType tmp_array(A.shape());
+  Subtract(A, B, tmp_array);
+  Square(tmp_array);
+
+  typename ArrayType::Type ret = Sum(tmp_array);
+  ret                          = Divide(ret, typename ArrayType::Type(A.shape()[0]));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/ml/loss_functions/mean_square_error.hpp]: Souldn't we divide by A.size rather than A.shape()[0] ?

File: libs/math/include/math/free_functions/ml/loss_functions/softmax_cross_entropy.hpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  //  of softmax layer)
+
+  auto      gt = ArgMax(y, 1);
+  ArrayType log_likelihood{1};
+  log_likelihood[0] = 0;
+
+  for (std::size_t idx = 0; idx < n_examples; ++idx)
+  {
+    sce_x.Set(idx, static_cast<std::size_t>(gt[idx]),
+              std::log(sce_x.At(idx, static_cast<std::size_t>(gt[idx]))));
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/ml/loss_functions/softmax_cross_entropy.hpp]: std::log will break with probably not compile with fixed point
I think Set should take a ArrayType::SizeType rather than std::size_t

File: libs/ml/include/ml/ops/activations/sigmoid.hpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    assert(inputs[0]->shape() == errorSignal->shape());
+
+    ArrayPtrType t = this->Forward(inputs);
+    fetch::math::Add(*errorSignal,
+                     fetch::math::Multiply(*t, fetch::math::Subtract(DataType(1), *t)),
+                     *errorSignal);
+    return {errorSignal};
+  }
+
+  static std::string Descriptor()
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/ops/activations/sigmoid.hpp]: Mismatch between the    static std::string Descriptor()  & static constexpr char const *DESCRIPTOR

File: libs/ml/include/ml/ops/loss_functions/cross_entropy.hpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    assert(inputs.size() == 2);
+    assert(inputs[0]->size() == inputs[1]->size());
+
+    typename ArrayType::Type n_classes = static_cast<typename ArrayType::Type>(inputs[1]->size());
+    ArrayPtrType             ret       = std::make_shared<ArrayType>(
+        fetch::math::Divide(fetch::math::Subtract(*inputs[0], *inputs[1]), n_classes));
+    return ret;
+  }
+
+  static constexpr char const *DESCRIPTOR = "CrossEntropy";
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/ops/loss_functions/cross_entropy.hpp:63]: Mismatch between the    static std::string Descriptor()  & static constexpr char const *DESCRIPTOR

File: libs/ml/include/ml/ops/loss_functions/softmax_cross_entropy.hpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+  {
+    assert(inputs.size() == 2);
+    assert(inputs[0]->size() == inputs[1]->size());
+
+    typename ArrayType::Type n_classes = static_cast<typename ArrayType::Type>(inputs[1]->size());
+    ArrayPtrType ret = std::make_shared<ArrayType>(fetch::math::Subtract(*inputs[0], *inputs[1]));
+    return ret;
+  }
+
+  static std::string Descriptor()
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/include/ml/ops/loss_functions/softmax_cross_entropy.hpp]: Mismatch between the    static std::string Descriptor()  & static constexpr char const *DESCRIPTOR

File: libs/ml/tests/ml/ops/activations/softmax.cpp [hunk 9]
```diff
@@ -58,7 +58,7 @@ TYPED_TEST(SoftmaxTest, backward_test)
   std::shared_ptr<TypeParam> gt    = std::make_shared<TypeParam>(8);
   std::vector<double>        dataInput({1, -2, 3, -4, 5, -6, 7, -8});
   std::vector<double>        errorInput({0, 0, 0, 0, 1, 0, 0, 0});
-  std::vector<double> gtInput({-2.5091e-04, -1.2492e-05, -1.8540e-03, -1.6906e-06, 1.0335e-01,
+  std::vector<double>        gtInput({-0.0002509, -1.2492e-05, -0.0018540, -1.6906e-06, 1.0335e-01,
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/tests/ml/ops/activations/softmax.cpp]: What happenned here ? 
Same inputs values but different results, was the previous test wrong ?

File: libs/ml/tests/ml/ops/activations/softmax.cpp [hunk 10]
```diff
@@ -69,6 +69,11 @@ TYPED_TEST(SoftmaxTest, backward_test)
   fetch::ml::ops::Softmax<TypeParam>      op;
   std::vector<std::shared_ptr<TypeParam>> prediction = op.Backward({data}, error);
 
+  for (std::size_t j = 0; j < prediction[0]->size(); ++j)
+  {
+    std::cout << "prediction[0]->At(j): " << prediction[0]->At(j) << std::endl;
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/tests/ml/ops/activations/softmax.cpp]: Leftover debug code

File: libs/ml/tests/ml/ops/loss_functions/cross_entropy.cpp [hunk 11]
```diff
@@ -99,8 +99,10 @@ TYPED_TEST(CrossEntropyTest, one_dimensional_forward_test)
     data1->Set(i, typename TypeParam::Type(logits[i]));
   }
 
-  fetch::ml::ops::CrossEntropyLayer<TypeParam> op;
-  ASSERT_FLOAT_EQ(float(op.Forward({data1, data2})), 2.3025851249694824);
+  fetch::ml::ops::CrossEntropy<TypeParam> op;
+  std::cout << "float(op.Forward({data1, data2})): " << float(op.Forward({data1, data2}))
```
Comment (pierre-wilmot-fetch-ai) [inline libs/ml/tests/ml/ops/loss_functions/cross_entropy.cpp]: Leftover debug code

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]
I checked the code, not the maths.
Please make sure to include more unit test coverage in a future PR.
This is especialy important since none of this number make intuive sense, so it's going to be a pain to debug at a later stage if something is wrong.

File: libs/math/include/math/free_functions/ml/loss_functions/l2_norm.hpp (hunk 3 above)
Comment (baykaner) [reply libs/math/include/math/free_functions/ml/loss_functions/l2_norm.hpp:35]: possibly - I'll do a bunch of unit tests in an upcoming PR to close (https://github.com/uvue-git/fetch-ledger/issues/711) and update as necessary at that time

File: libs/math/include/math/free_functions/ml/activation_functions/relu.hpp (hunk 2 above)
Comment (baykaner) [reply libs/math/include/math/free_functions/ml/activation_functions/relu.hpp:47]: since that solution doesn't work for shapeless array I'll remove shapeless array from relu test

File: libs/math/include/math/free_functions/ml/loss_functions/mean_square_error.hpp (hunk 4 above)
Comment (baykaner) [reply libs/math/include/math/free_functions/ml/loss_functions/mean_square_error.hpp]: that's probably true - i think this was leftover from assuming shape()[0] would always be the batch dimension. We will still update the loss functions as we implement batching

File: libs/math/include/math/free_functions/ml/loss_functions/softmax_cross_entropy.hpp (hunk 5 above)
Comment (baykaner) [reply libs/math/include/math/free_functions/ml/loss_functions/softmax_cross_entropy.hpp]: next PR to implement unit tests and fix failing fixed point implementations where necessary.
you're right about std::size_t

File: libs/ml/tests/ml/ops/activations/softmax.cpp (hunk 9 above)
Comment (baykaner) [reply libs/ml/tests/ml/ops/activations/softmax.cpp]: no just an error where I was debugging the wrong test by accident. fixed

File: libs/math/include/math/free_functions/fundamental_operators.hpp [hunk 12]
```diff
@@ -732,6 +650,104 @@ meta::IfIsArithmetic<S, void> Multiply(S const &scalar1, S const &scalar2, S &re
...
+  return ret;
+}
+template <typename S>
+meta::IfIsArithmetic<S, S> Multiply(S const &scalar1, S const &scalar2)
+{
+  S ret;
+  Multiply(scalar1, scalar2, ret);
+  return ret;
+}
+
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/fundamental_operators.hpp:750]: That's a lot of  Multiply 🙄

File: libs/math/include/math/free_functions/ml/activation_functions/relu.hpp [hunk 13]
```diff
@@ -41,8 +41,7 @@ void Relu(ArrayType const &t, ArrayType &ret)
 template <typename ArrayType>
 ArrayType Relu(ArrayType &t)
 {
-  ArrayType ret;
-  ret.Copy(t);
+  ArrayType ret(t.shape());
   Relu(ret, ret);
```
Comment (pierre-wilmot-fetch-ai) [inline libs/math/include/math/free_functions/ml/activation_functions/relu.hpp]: Relu(t, ret);

**Review by pierre-wilmot-fetch-ai — CHANGES_REQUESTED** [review]

**Review by pierre-wilmot-fetch-ai — APPROVED** [review]
